src/extensions/default/NavigationAndHistory/main.js

```diff
@@ -0,0 +1,402 @@
+                                                    paneId: contextData.paneId }).done(function () {
+            EditorManager.getActiveEditor().setCursorPos(contextData.cursor);
+            EditorManager.getActiveEditor().centerOnCursor();
+        });
+    }
+    
+    /**
+     * Creates an entry for MROF list
+     * @private
+     * @param {String} path - full path of a doc
+     * @param {String} pane - the pane holding the editor for the doc
+     * @param {Object} cursorPos - current cursor position
+     * @return {Object} a frame containing file path, pane and last known cursor
+     */
+    function _makeMROFListEntry(path, pane, cursorPos) {
+        return {
+            file: path,
+            paneId: pane,
+            cursor: cursorPos
+        };
+    }
+    
+    /**
+     * Determines if a file is dirty
+     * @private
+     * @param {!File} file - file to test
+     * @return {boolean} true if the file is dirty, false otherwise
+     */
+    function _isOpenAndDirty(file) {
+        // working set item might never have been opened; if so, then it's definitely not dirty
+        var docIfOpen = DocumentManager.getOpenDocumentForPath(file.fullPath);
+        return (docIfOpen && docIfOpen.isDirty);
+    }
+    
+    /** Returns a 'context' object for getting/setting project-specific preferences */
+    function _getPrefsContext() {
+        var projectRoot = ProjectManager.getProjectRoot();
+        return { location : { scope: "user", layer: "project", layerID: projectRoot && projectRoot.fullPath } };
+    }
+    
+    
+    function _checkExt(entry, index) {
+        var deferred = new $.Deferred(),
+            fileEntry = FileSystem.getFileForPath(entry.file);
+
+        fileEntry.exists(function (err, exists) {
+            if (!err && exists) {
+                deferred.resolve();
+            } else {
+                _mrofList[index] = null;
+                deferred.reject();
+            }
+        });
+
+        return deferred.promise();
+    }
+    
+    /**
+     * Checks whether entries in MROF list actually exists in fileSystem to prevent access to deleted files 
+     * @private
+     */
+    function _syncWithFileSystem() {
+        _mrofList = _mrofList.filter(function (e) {return e; });
+        Async.doSequentially(_mrofList, _checkExt, false);
+        _mrofList = _mrofList.filter(function (e) {return e; });
+    }
+    
+    /**
+     * Hides the current MROF list if visible
+     * @private
+     */
+    function _hideMROFList() {
+        if ($mrofContainer) {
+            $mrofContainer.remove();
+            $mrofContainer = null;
+            $currentContext = null;
+            if (EditorManager.getActiveEditor()) {
+                EditorManager.getActiveEditor().focus();
+            }
+            hideTimeoutVar = null;
+        }
+    }
+    
+    /**
+     * Shows the current MROF list
+     * @private
+     */
+    function _createMROFDisplayList() {
```

Comment:
This function spans almost 100 lines. Most likely could use some refactoring :pencil2: